lib/tcp_connection.ex

```diff
@@ -0,0 +1,132 @@
+defmodule ElixirSonicClient.TcpConnection do
+  @moduledoc """
+  This is the TcpConnection module, responsible to send and receive calls.
+  """
+  use Connection
+
+  @doc """
+  Starts connection with the tcp server.
+
+  Returns `{:ok, conn}`.
+
+  ## Examples
+
+      iex> ElixirSonicClient.TcpConnection.start_link(127.0.0.1, 1491, [])
+      {:ok, #PID<0.198.0>}
+  """
+  def start_link(host, port, opts, timeout \\ 5000) do
+    Connection.start_link(__MODULE__, {host, port, opts, timeout})
+  end
+
+  @doc """
+  Sends a message to the tcp server.
+
+  Returns `:ok`.
+
+  ## Examples
+
+      iex> ElixirSonicClient.TcpConnection.send(conn, "start search password")
+      :ok
+  """
+  def send(conn, data), do: Connection.call(conn, {:send, data <> "\n"})
+
+  @doc """
+  Receives message from the tcp server.
+
+  Returns `{:ok, response}`.
+
+  ## Examples
+
+      iex> ElixirSonicClient.TcpConnection.recv(conn)
+      {:ok, 'CONNECTED <sonic-server v1.3.0>\r\n'}
+  """
+  def recv(conn, bytes \\ 0, timeout \\ 3000) do
+    complete_response(conn, bytes, timeout)
+  end
+
+  defp complete_response(responses \\ [], conn, bytes, timeout) do
+    {:ok, response} = Connection.call(conn, {:recv, bytes, timeout})
```

Comment:
So, does this mean that although `Connection.call` is synchronous (it is, right?) it may return for `:recv` with a value for `response` that does not represent a complete Sonic protocol messsage? That is, since this is a basic TCP connection,  is it the case that it is not aware of the contents of what's being sent through it and reading it may return after an arbitrary TCP segment payload size?  Just trying to make sure that I understand what is going on, here.  🙂

Reply:
Yes `call` is synchronous. So for what I understood response does not represent the sonic protocol message until it has a `\n` the same when we send a message, sonic will understand that it's finished when the message has `\n`.

Reply:
Yeah ... I see.